pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,201 @@
+	// otherwise write it into the file and queue
+	var backupQueue = list.New()
+	for queue.Len() > 0 {
+		e := queue.Front().Value.(zapcore.Entry)
+		if e.Message != newErrorLog.Message {
+			n, err = errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				return byteNum, err
+			}
+			byteNum += n
+			backupQueue.PushBack(e)
+		}
+		queue.Remove(queue.Front())
+	}
+	backupQueue.PushBack(newErrorLog)
+	queue = backupQueue
+	n, err = errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+		newErrorLog.Time, newErrorLog.Level, newErrorLog.Caller, newErrorLog.Message)))
+	if err != nil {
+		return byteNum, err
+	}
+	byteNum += n
+	if byteNum >= kilobyte {
+		byteNum, err = rotate(byteNum, errorLogfile)
+		if err != nil {
+			return byteNum, err
+		}
+	}
+	return byteNum, err
+}
+
+// rotate func could delete the old logs to make sure error log file less than 1 KB
+func rotate(byteNum int, errorLogfile string) (n int, err error) {
+	file, err := os.OpenFile(errorLogfile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC|os.O_APPEND, 0777)
+	if err != nil {
+		return byteNum, err
+	}
+	err = file.Truncate(0)
+	if err != nil {
+		return byteNum, err
+	}
+	errorNum := queue.Len()
+	for i := 0; i < errorNum; i++ {
+		currentError := queue.Front().Value.(zapcore.Entry)
+		currentLog := []byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+			currentError.Time, currentError.Level, currentError.Caller, currentError.Message))
+		if byteNum > kilobyte {
+			byteNum -= len(currentLog)
+			queue.Remove(queue.Front())
+		} else {
+			_, err = file.Write(currentLog)
+			if err != nil {
+				return byteNum, err
+			}
+		}
+
+	}
+	return byteNum, err
+}
+
+var RotateError = make(chan error)
+
+// RotateErrorLog func help select the different case between either write error log into file or
+// rotate and delete old logs in file based on TTL
+func RotateErrorLog() {
+	defer ticker.Stop()
+	for {
+		select {
+		// when new error add in the channel, this case will write it into file
+		case e := <- errorLogChannel:
+			_, err := WriteError(e, errorLogfile)
+			if err != nil {
+				RotateError <- err
+			}
+		// this case will run every x mins (x is configurable) to delete the error which already expired
+		case <- ticker.C:
+			currentTime := time.Now()
+			file, _ := os.OpenFile(errorLogfile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC|os.O_APPEND, 0777)
+			_ = file.Truncate(0)
+			for queue.Len() > 0 {
+				e := queue.Front().Value.(zapcore.Entry)
+				if e.Time.Add(time.Duration(errorLogMaxAge) * time.Hour).After(currentTime) {
+					_, err := file.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+						e.Time, e.Level, e.Caller, e.Message)))
+					if err != nil {
+						RotateError <- err
+					}
+				} else {
+					queue.Remove(queue.Front())
+				}
+			}
```

Comment:
the error will contains a timestamp, and i think this goroutine will keep running with the AOC, so if there is some error shows up, it will get this error channel, which doesn't related to the rotate case